### PR TITLE
SMP patch series

### PR DESCRIPTION
**Pull request by andyross**
This is the full, cleaned up and apparently working SMPification series for APL/up2.  It includes the linker work for cacheability, a few minor bug fixes and workarounds, and the configuration logic to make SMP the default (but not in the SOF app itself, because I don't have a full test rig for that and am afraid of breaking something).

It also includes the printk hook, which isn't strictly related but allows for synchronous early logging without the log subsystem enabled.  That makes my world MUCH easier.  It works the way you expect: disable LOG=y and your printk() output goes out synchronously to the same buffer and can be read via the same tools.

Also, it removes some SOF SMP code that had been added AFAICT mistakenly to the Zephyr tree.  Zephyr doesn't use that stuff anywhere.  I'll submit a separate patch to add it back to the SOF Zephyr wrapper API.  But I do note that SOF wasn't building with SMP in tree anyway, so it's likely there is other bitrot to fix.

With these applied, pretty much the full Zephyr test suite runs correctly with or without SMP enabled (at least all the tests I tried -- we don't have a proper sanitycheck integration so I have to do them manually).   But while the SOF rig itself builds and runs apparently identically, I still don't have working audio output on my board so I'm not particularly confident I haven't broken anything.

Finally: I'm too chicken to merge this on my own.  If the folks reviewing (and ESPECIALLY those with working audio output!) could actually try this code and report success/failure I'd be really appreciative.

### CODE REVIEW

**Comment by andyross** [comment]
Took a look at enabling SMP on SOF proper.  The patch below works to make it build and isolates the dependencies cleanly, but there's some significant untangling to do here.  The root problem is that when CONFIG_SMP is true, SOF wants to do a ton of low level CPU initialization and direction that Zephyr has already done, including its own IDC driver, etc...

I don't know enough about the SOF architecture yet, but I suspect what really needs to happen is that we want to build the SOF code in what to it looks like **non** SMP mode.  That's the point of symmetric multiprocessing -- user code doesn't care and just gets to run when a CPU is available.  We'd need to find all the spots where it's doing synchronization stuff like cache flushing and keep those, but leave all the architecture uniprocessor.  I think.  @lgirdwood et. al. will need to chime in on the best way forward here.

```patch
diff --git a/src/drivers/intel/cavs/CMakeLists.txt b/src/drivers/intel/cavs/CMakeLists.txt
index 6ee55286..ec645fae 100644
--- a/src/drivers/intel/cavs/CMakeLists.txt
+++ b/src/drivers/intel/cavs/CMakeLists.txt
@@ -33,6 +33,7 @@ if(CONFIG_CAVS_ALH)
        add_local_sources(sof alh.c)
 endif()
 
-if(CONFIG_SMP)
+# Build IDC driver only when not building in Zephyr
+if(CONFIG_SMP AND NOT CONFIG_SOC_FAMILY_INTEL_ADSP)
        add_local_sources(sof idc.c)
 endif()
diff --git a/zephyr/wrapper.c b/zephyr/wrapper.c
index 5bd7c9e4..979792c0 100644
--- a/zephyr/wrapper.c
+++ b/zephyr/wrapper.c
@@ -478,3 +478,33 @@ int task_main_start(void)
 
        return 0;
 }
+
+void cpu_power_down_core(void)
+{
+       /* TODO: this is linked against, but never called?  Zephyr
+        * currently dosn't have support for offlining CPUs.
+        */
+}
+
+void cpu_alloc_core_context(int core)
+{
+       /* TODO: seems like this is called, but the "context"
+        * allocated is never used externally?
+        */
+}
+
+int arch_cpu_is_core_enabled(int id)
+{
+       /* TODO */
+       return id >= 0 && id < CONFIG_MP_NUM_CPUS;
+}
+
+int idc_send_msg(struct idc_msg *msg, uint32_t mode)
+{
+       /* TODO */
+}
+
+int idc_init(void)
+{
+       /* No-op, handled by Zephyr cavs_idc driver */
+}
```

File: soc/xtensa/intel_adsp/cavs_v15/linker.ld [hunk 1]
```diff
@@ -422,18 +453,25 @@ SECTIONS
...
+ * wants to use the cache
+  */
+#undef RAMABLE_REGION
+#undef ROMABLE_REGION
+#define RAMABLE_REGION ucram :ucram_phdr
+#define ROMABLE_REGION ucram :ucram_phdr
 #include <linker/common-ram.ld>
 
-  .bss (NOLOAD) : ALIGN(4096)
+  /* ANDY TEST */
```
Comment (dcpleung) [inline soc/xtensa/intel_adsp/cavs_v15/linker.ld]: remove?

**Comment by lgirdwood** [comment]
@andyross does this stiil build and boot on qemu for cav18, cavs20 targets and build for cavs25 ?

**Comment by andyross** [comment]
Oops, I'm not quite done.  This drops the patch that removed the cacheattr setings when SMP (and added a flush on thread creation), so the successes are sorta faked.  Will get that added today.  And I realize that I'm missing a flush on switch and interrupt context changes that I need to add (even though they don't seem to be exercised).  Will have that ready today.

As far as other targets: it builds on those still (though I didn't smoke the final version).  I'm not running qemu yet, though I suppose I should be.

**Comment by andyross** [comment]
OK, **now** it works.  The subtleties involved in doing proper cache flushing of thread stacks turned out to be, kinda mindbendingly difficult.  But this series now runs more or less everything.  Some notes:

* One difficulty is in the k_poll API, which isn't "coherence safe" at the moment.   A few tests turn this on and you have to manually disable it (and the test cases that use poll) to get them to pass.

* tests/kernel/common won't work with the log backend, I haven't had time to debug.  You need to set CONFIG_LOG=n and use the printk backend to get it to produce output.

As before, this **appears** to behave identically when I run the samples/sof code under the proper driver and start an output stream.  But as I don't actually have audio working yet I'm not super confident that nothing is broken.  This definitely needs validation from someone with a working rig.

**Comment by lgirdwood** [comment]
@andyross will give it a try on real HW today :)

**Pull request merged**